io: Remove blanket trait impls

tcharding opened the pull request:
Remove the blanket impls of `Read`, `BufRead`, and `Write`. Replace them with just the impls required to get `rust-bitcoin` building.

Note, the `TcpStream` stuff is used in `examples/handshake.rs`.

Fix: #2432

tcharding commented:
@apoelstra I did this instead of trying the trait idea we discussed - my aim is to make forward progress, by tooth and nail. This makes the crate non-broken and as such releasable. It may be missing functionality which seems reasonable for a `v0.1.0`.

tcharding commented:
Just noticed I missed doing everything in https://github.com/rust-bitcoin/rust-bitcoin/issues/2432#issuecomment-1928116934

tcharding commented:
@TheBlueMatt is this PR going to rug you?

Kixunil commented on io/src/lib.rs:
```diff
@@ -143,35 +143,32 @@ impl<'a, R: BufRead + ?Sized> BufRead for Take<'a, R> {
-#[cfg(not(feature = "std"))]
-impl Read for &[u8] {
+#[cfg(feature = "std")]
+impl Read for std::io::BufReader<std::net::TcpStream> {
```
This needs to be `T: std::io::Read`, no point implementing it for `TcpStream` only.

tcharding replied:
I had that at first but we are specifically trying to remove the blanket impls so I did this?

Kixunil replied:
`impl<T: std::io::Read> for std::io::BufReader<T>` is not really a blanket impl. Or at least not in the sense that would break things.

apoelstra replied:
Right -- `BufReader` is a "concrete type" even though it's parameterized.

Kixunil commented on io/src/lib.rs:
```diff
@@ -181,6 +178,15 @@ impl BufRead for &[u8] {
 }
 
+#[cfg(feature = "std")]
+impl BufRead for std::io::BufReader<std::net::TcpStream> {
```
Generic

Kixunil commented on io/src/lib.rs:
```diff
@@ -296,18 +289,6 @@ impl Write for Sink {
-
-    #[inline]
-    fn flush(&mut self) -> std::io::Result<()> { Ok(()) }
-}
```
We can and should impl it for `BufWriter`.

tcharding replied:
I'm at the point that I think we should just release this with whatever works for `rust-bitcoin` as it is. I don't know what is the plan for this crate in the future so I don't want to add anything that is not used. I just want to use the crate already in `hex` so I can fix stuff in `rust-bitcoin` that needs it.

Kixunil replied:
I don't think we should release total crap just because "it works" in our library. This is literally a few lines that will help with super common reasonable cases downstream. Without these actual IO would be insanely annoying to use.

tcharding commented:
I've implemented for `BufWriter` as requested. Also `Write` for `TcpStream` since we do `Read`. Also I've added generics to the `BufRead` and `BufWrite` impls.

tcharding commented:
Please merge #2352 before this one.

tcharding commented:
Rebased and improved commit message to mention the macro changes. No other code changes.

apoelstra reviewed (approved):
ACK 249459b728b77197bda26c5bf4da28b1cef2644d thanks for doing this!

tcharding commented:
No sweat, lets get this first release done - `rust-bitcoin` devs, gotta ship.

Kixunil commented:
@apoelstra I suggest let's wait for Tobin to address my comments before merging. I mainly ACK so you can merge if Tobin doesn't want to change it because I think it's acceptable.

tcharding commented:
Happy to change, all suggestions seem good to me. I'll hack it up.

tcharding commented on io/src/lib.rs:
```diff
@@ -151,35 +151,28 @@ impl<'a, R: BufRead + ?Sized> BufRead for Take<'a, R> {
-    #[inline]
-    fn fill_buf(&mut self) -> Result<&[u8]> { Ok(std::io::BufRead::fill_buf(self)?) }
-
+impl Read for std::net::TcpStream {
```
Ah yes, and I'd expect anyone using a TCP stream to use the buffered reader - removed.

tcharding replied:
Removed `Write for TcpStream` also for the same reason.

tcharding commented on io/src/lib.rs:
```diff
@@ -316,18 +323,6 @@ impl Write for Sink {
 }
 
-#[cfg(feature = "std")]
-impl std::io::Write for Sink {
```
I'm confused. You mean our `Write`, right? That is implemented at the moment.

Kixunil replied:
Yes, that. Good.

tcharding commented on io/src/lib.rs:
```diff
@@ -151,35 +151,28 @@ impl<'a, R: BufRead + ?Sized> BufRead for Take<'a, R> {
     }
```
Ha, thanks.

tcharding commented:
All done except the sink stuff. I think the sink is correct, it currently is:
```rust
/// A sink to which all writes succeed. See [`std::io::Sink`] for more info.
///
/// Created using `io::sink()`.
pub struct Sink;

impl Write for Sink {
    #[inline]
    fn write(&mut self, buf: &[u8]) -> Result<usize> { Ok(buf.len()) }

    #[inline]
    fn write_all(&mut self, _: &[u8]) -> Result<()> { Ok(()) }

    #[inline]
    fn flush(&mut self) -> Result<()> { Ok(()) }
}

/// Returns a sink to which all writes succeed. See [`std::io::sink`] for more info.
#[inline]
pub fn sink() -> Sink { Sink }
```

apoelstra commented:
92cb9026111924b3af44300afdce8621d9b8353a looks good except that it removes the `std::io::Write` impl for our `Sink` ... but I think we can/should keep this impl.

tcharding commented:
Woops, will put it back in. Hard to get good help huh.

tcharding commented:
Add the impl back in and made the impl of our `Write` unconditional i.e., always implement our `Write` and implement `std::io::Write` if "std" is enabled. Same as what the macros do.

Kixunil reviewed (approved):
ACK db888fa4cc3631634ec43aa992807d9aa8f9a106

Kixunil commented:
FTR we don't actually use the `Read` trait since we use `BufRead` and don't provide our own `BufReader`. So we could actually remove the `Read` trait and move helpers on top of `BufRead`. :) Not sure if it's worth the effort though.

tcharding commented:
Fair points, lets leave it as is for now and see where the `io` crate goes. At the moment it is doing its job.

apoelstra reviewed (approved):
ACK db888fa4cc3631634ec43aa992807d9aa8f9a106
